caffe2 - Util to cleanup external inputs and outputs from a NetDef

duc0 opened the pull request:
Summary:
Add a util method to cleanup external inputs and outputs from a NetDef

The following conditions will be met after the modification
- No duplicate external inputs
- No duplicate external outputs
- Going through list of ops in order, all op inputs must be outputs
from other ops, or registered as external inputs.
- All external outputs must be outputs of some operators.

Differential Revision: D14528589

ZolotukhinM commented on caffe2/utils/proto_utils.cc:
```diff
@@ -558,4 +558,55 @@ C10_EXPORT Argument* GetMutableArgument(
+    net->add_external_input(input);
+  }
+
+  std::set<std::string> allOutputs;
```
Should we also include external inputs here?

duc0 committed:
caffe2 - Util to cleanup external inputs and outputs from a NetDef (#18194)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18194

Add a util method to cleanup external inputs and outputs from a NetDef

The following conditions will be met after the modification
- No duplicate external inputs
- No duplicate external outputs
- Going through list of ops in order, all op inputs must be outputs
from other ops, or registered as external inputs.
- All external outputs must be outputs of some operators.

Reviewed By: ZolotukhinM

Differential Revision: D14528589

fbshipit-source-id: 20820ff2afacacef74bfdd60b48436d8d175d1e6
